File: config-differ.py
#!/usr/bin/env python3
"""
Config Differ - 智能配置文件对比工具

支持多种格式:
- YAML/YML
- JSON
- INI/CONF
- 纯文本

功能:
- 自动识别文件类型
- 结构化对比（YAML/JSON）
- 忽略注释和空行
- 彩色输出
- 支持远程文件
- 多环境批量对比

作者: DevOps Team
"""

import argparse
import json
import yaml
import sys
import os
import subprocess
import tempfile
from pathlib import Path
from typing import Dict, List, Tuple, Optional
from difflib import unified_diff, HtmlDiff
from colorama import init, Fore, Style
import configparser

init(autoreset=True)


class ConfigDiffer:
    def __init__(self, ignore_comments=True, ignore_blank=True, 
                 ignore_order=False, context_lines=3):
        self.ignore_comments = ignore_comments
        self.ignore_blank = ignore_blank
        self.ignore_order = ignore_order
        self.context_lines = context_lines
    
    def detect_type(self, filepath: str) -> str:
        """检测文件类型"""
        ext = Path(filepath).suffix.lower()
        
        type_map = {
            '.yaml': 'yaml',
            '.yml': 'yaml',
            '.json': 'json',
            '.ini': 'ini',
            '.conf': 'ini',
            '.cfg': 'ini',
            '.properties': 'properties',
            '.env': 'env',
        }
        
        return type_map.get(ext, 'text')
    
    def read_file(self, filepath: str) -> str:
        """读取文件内容"""
        if filepath.startswith('http://') or filepath.startswith('https://'):
            # 远程文件
            import urllib.request
            with urllib.request.urlopen(filepath) as response:
                return response.read().decode('utf-8')
        
        if ':' in filepath and not filepath.startswith('/'):
            # SSH远程文件 (host:path)
            host, path = filepath.split(':', 1)
            result = subprocess.run(
                ['ssh', host, f'cat {path}'],
                capture_output=True,
                text=True
            )
            if result.returncode != 0:
                raise Exception(f"无法读取远程文件: {result.stderr}")
            return result.stdout
        
        # 本地文件
        with open(filepath, 'r', encoding='utf-8') as f:
            return f.read()
    
    def normalize_text(self, content: str, file_type: str) -> List[str]:
        """标准化文本（去除注释、空行等）"""
        lines = content.split('\n')
        result = []
        
        for line in lines:
            # 去除空行
            if self.ignore_blank and not line.strip():
                continue
            
            # 去除注释
            if self.ignore_comments:
                if file_type in ['yaml', 'ini', 'properties', 'env']:
                    # #注释
                    if line.strip().startswith('#'):
                        continue
                    # 行尾注释
                    if '#' in line:
                        line = line.split('#')[0].rstrip()
                
                elif file_type == 'json':
                    # JSON标准不支持注释，但有些工具支持
                    if line.strip().startswith('//'):
                        continue
            
            result.append(line)
        
        return result
    
    def parse_yaml(self, content: str) -> dict:
        """解析YAML并排序（可选）"""
        data = yaml.safe_load(content)
        if self.ignore_order and isinstance(data, dict):
            return self._sort_dict_recursive(data)
        return data
    
    def parse_json(self, content: str) -> dict:
        """解析JSON并排序（可选）"""
        data = json.loads(content)
        if self.ignore_order and isinstance(data, dict):
            return self._sort_dict_recursive(data)
        return data
    
    def _sort_dict_recursive(self, obj):
        """递归排序字典"""
        if isinstance(obj, dict):
            return {k: self._sort_dict_recursive(v) for k, v in sorted(obj.items())}
        elif isinstance(obj, list):
            return [self._sort_dict_recursive(item) for item in obj]
        return obj
    
    def diff_structured(self, file1: str, file2: str, file_type: str) -> str:
        """结构化对比（YAML/JSON）"""
        content1 = self.read_file(file1)
        content2 = self.read_file(file2)
        
        try:
            if file_type == 'yaml':
                data1 = self.parse_yaml(content1)
                data2 = self.parse_yaml(content2)
                # 转为格式化YAML字符串
                str1 = yaml.dump(data1, default_flow_style=False, sort_keys=True)
                str2 = yaml.dump(data2, default_flow_style=False, sort_keys=True)
            else:  # json
                data1 = self.parse_json(content1)
                data2 = self.parse_json(content2)
                # 转为格式化JSON字符串
                str1 = json.dumps(data1, indent=2, sort_keys=True)
                str2 = json.dumps(data2, indent=2, sort_keys=True)
            
            return self._unified_diff(str1, str2, file1, file2)
        
        except Exception as e:
            print(f"{Fore.YELLOW}警告: 结构化解析失败，回退到文本对比: {e}")
            return self.diff_text(file1, file2, file_type)
    
    def diff_text(self, file1: str, file2: str, file_type: str) -> str:
        """文本对比"""
        content1 = self.read_file(file1)
        content2 = self.read_file(file2)
        
        lines1 = self.normalize_text(content1, file_type)
        lines2 = self.normalize_text(content2, file_type)
        
        return self._unified_diff('\n'.join(lines1), '\n'.join(lines2), file1, file2)
    
    def _unified_diff(self, str1: str, str2: str, file1: str, file2: str) -> str:
        """生成unified diff"""
        lines1 = str1.splitlines(keepends=True)
        lines2 = str2.splitlines(keepends=True)
        
        diff = unified_diff(
            lines1,
            lines2,
            fromfile=file1,
            tofile=file2,
            n=self.context_lines
        )
        
        return self._colorize_diff(diff)
    
    def _colorize_diff(self, diff) -> str:
        """彩色化diff输出"""
        result = []
        for line in diff:
            if line.startswith('---') or line.startswith('+++'):
                result.append(f"{Fore.CYAN}{Style.BRIGHT}{line}{Style.RESET_ALL}")
            elif line.startswith('@@'):
                result.append(f"{Fore.MAGENTA}{line}{Style.RESET_ALL}")
            elif line.startswith('+'):
                result.append(f"{Fore.GREEN}{line}{Style.RESET_ALL}")
            elif line.startswith('-'):
                result.append(f"{Fore.RED}{line}{Style.RESET_ALL}")
            else:
                result.append(line)
        
        return ''.join(result)
    
    def compare(self, file1: str, file2: str) -> Tuple[bool, str]:
        """对比两个文件"""
        file_type = self.detect_type(file1)
        
        print(f"{Fore.CYAN}文件类型: {file_type}")
        print(f"{Fore.CYAN}文件1: {file1}")
        print(f"{Fore.CYAN}文件2: {file2}")
        print(f"{Fore.CYAN}{'='*60}{Style.RESET_ALL}\n")
        
        if file_type in ['yaml', 'json']:
            diff_output = self.diff_structured(file1, file2, file_type)
        else:
            diff_output = self.diff_text(file1, file2, file_type)
        
        if not diff_output.strip():
            return True, f"{Fore.GREEN}✓ 文件完全相同{Style.RESET_ALL}"
        else:
            return False, diff_output
    
    def compare_dirs(self, dir1: str, dir2: str, pattern: str = "*") -> Dict[str, str]:
        """对比两个目录"""
        from pathlib import Path
        import fnmatch
        
        results = {}
        path1 = Path(dir1)
        path2 = Path(dir2)
        
        # 获取所有匹配文件
        files1 = set()
        for f in path1.rglob(pattern):
            if f.is_file():
                files1.add(f.relative_to(path1))
        
        files2 = set()
        for f in path2.rglob(pattern):
            if f.is_file():
                files2.add(f.relative_to(path2))
        
        # 只在dir1中
        only_in_1 = files1 - files2
        if only_in_1:
            print(f"\n{Fore.YELLOW}只在 {dir1} 中:{Style.RESET_ALL}")
            for f in sorted(only_in_1):
                print(f"  {f}")
        
        # 只在dir2中
        only_in_2 = files2 - files1
        if only_in_2:
            print(f"\n{Fore.YELLOW}只在 {dir2} 中:{Style.RESET_ALL}")
            for f in sorted(only_in_2):
                print(f"  {f}")
        
        # 两边都有的文件
        common = files1 & files2
        
        print(f"\n{Fore.CYAN}对比共有文件 ({len(common)} 个):{Style.RESET_ALL}\n")
        
        for rel_path in sorted(common):
            f1 = path1 / rel_path
            f2 = path2 / rel_path
            
            print(f"\n{Fore.CYAN}{'='*60}")
            print(f"对比: {rel_path}")
            print(f"{'='*60}{Style.RESET_ALL}")
            
            same, diff = self.compare(str(f1), str(f2))
            results[str(rel_path)] = 'same' if same else 'different'
            
            if same:
                print(diff)
            else:
                print(diff)
        
        return results


def main():
    parser = argparse.ArgumentParser(
        description='智能配置文件对比工具',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:

  # 对比两个本地文件
  %(prog)s prod.yaml test.yaml

  # 对比本地和远程
  %(prog)s local.conf server1:/etc/nginx/nginx.conf

  # 对比两个目录
  %(prog)s -d /etc/nginx/prod /etc/nginx/test

  # YAML对比（忽略顺序）
  %(prog)s --ignore-order prod.yaml test.yaml

  # 包含注释和空行
  %(prog)s --no-ignore-comments prod.conf test.conf

  # 指定文件模式
  %(prog)s -d /etc/prod /etc/test -p "*.conf"
        """
    )
    
    parser.add_argument('file1', nargs='?', help='第一个文件或目录')
    parser.add_argument('file2', nargs='?', help='第二个文件或目录')
    
    parser.add_argument('-d', '--directory', action='store_true',
                       help='对比目录而非文件')
    parser.add_argument('-p', '--pattern', default='*',
                       help='目录对比时的文件模式 (默认: *)')
    
    parser.add_argument('--ignore-order', action='store_true',
                       help='忽略YAML/JSON的键顺序')
    parser.add_argument('--no-ignore-comments', action='store_true',
                       help='不忽略注释')
    parser.add_argument('--no-ignore-blank', action='store_true',
                       help='不忽略空行')
    
    parser.add_argument('-c', '--context', type=int, default=3,
                       help='上下文行数 (默认: 3)')
    
    parser.add_argument('--html', metavar='OUTPUT',
                       help='输出HTML格式报告')
    
    args = parser.parse_args()
    
    if not args.file1 or not args.file2:
        parser.print_help()
        sys.exit(1)
    
    differ = ConfigDiffer(
        ignore_comments=not args.no_ignore_comments,
        ignore_blank=not args.no_ignore_blank,
        ignore_order=args.ignore_order,
        context_lines=args.context
    )
    
    try:
        if args.directory:
            results = differ.compare_dirs(args.file1, args.file2, args.pattern)
            
            # 统计
            same = sum(1 for v in results.values() if v == 'same')
            diff = sum(1 for v in results.values() if v == 'different')
            
            print(f"\n{Fore.CYAN}{'='*60}")
            print(f"汇总")
            print(f"{'='*60}{Style.RESET_ALL}")
            print(f"{Fore.GREEN}相同: {same}")
            print(f"{Fore.RED}差异: {diff}")
            print(f"总计: {same + diff}")
        else:
            same, diff = differ.compare(args.file1, args.file2)
            print(diff)
            sys.exit(0 if same else 1)
    
    except Exception as e:
        print(f"{Fore.RED}错误: {e}{Style.RESET_ALL}", file=sys.stderr)
        sys.exit(2)


if __name__ == '__main__':
    main()

File: config-differ.go
package main

/*
Config Differ - Go版本
高性能配置文件对比工具

编译:
  go mod init config-differ
  go get gopkg.in/yaml.v3
  go build -o config-differ

使用:
  ./config-differ file1.yaml file2.yaml
  ./config-differ --ignore-order prod.json test.json
  ./config-differ -d /etc/nginx/prod /etc/nginx/test
*/

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/pmezard/go-difflib/difflib"
	"gopkg.in/yaml.v3"
)

// ANSI颜色
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorCyan   = "\033[36m"
	ColorBold   = "\033[1m"
)

type Config struct {
	IgnoreComments bool
	IgnoreBlank    bool
	IgnoreOrder    bool
	Context        int
}

func main() {
	var (
		file1          string
		file2          string
		isDir          bool
		pattern        string
		ignoreOrder    bool
		noIgnoreComments bool
		noIgnoreBlank  bool
		context        int
	)

	flag.BoolVar(&isDir, "d", false, "对比目录")
	flag.BoolVar(&isDir, "directory", false, "对比目录")
	flag.StringVar(&pattern, "p", "*", "文件模式")
	flag.StringVar(&pattern, "pattern", "*", "文件模式")
	flag.BoolVar(&ignoreOrder, "ignore-order", false, "忽略键顺序")
	flag.BoolVar(&noIgnoreComments, "no-ignore-comments", false, "不忽略注释")
	flag.BoolVar(&noIgnoreBlank, "no-ignore-blank", false, "不忽略空行")
	flag.IntVar(&context, "c", 3, "上下文行数")
	flag.IntVar(&context, "context", 3, "上下文行数")

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "用法: %s [选项] <file1> <file2>\n\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "选项:\n")
		flag.PrintDefaults()
		fmt.Fprintf(os.Stderr, "\n示例:\n")
		fmt.Fprintf(os.Stderr, "  %s prod.yaml test.yaml\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s --ignore-order prod.json test.json\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s -d /etc/nginx/prod /etc/nginx/test\n", os.Args[0])
	}

	flag.Parse()

	args := flag.Args()
	if len(args) != 2 {
		flag.Usage()
		os.Exit(1)
	}

	file1 = args[0]
	file2 = args[1]

	config := Config{
		IgnoreComments: !noIgnoreComments,
		IgnoreBlank:    !noIgnoreBlank,
		IgnoreOrder:    ignoreOrder,
		Context:        context,
	}

	if isDir {
		compareDirs(file1, file2, pattern, config)
	} else {
		same, diff := compareFiles(file1, file2, config)
		fmt.Println(diff)
		if !same {
			os.Exit(1)
		}
	}
}

func detectType(filename string) string {
	ext := strings.ToLower(filepath.Ext(filename))
	switch ext {
	case ".yaml", ".yml":
		return "yaml"
	case ".json":
		return "json"
	case ".ini", ".conf", ".cfg":
		return "ini"
	default:
		return "text"
	}
}

func readFile(filename string) (string, error) {
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func normalizeText(content string, fileType string, config Config) []string {
	lines := strings.Split(content, "\n")
	var result []string

	for _, line := range lines {
		// 去除空行
		if config.IgnoreBlank && strings.TrimSpace(line) == "" {
			continue
		}

		// 去除注释
		if config.IgnoreComments {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "#") {
				continue
			}
			// 去除行尾注释
			if idx := strings.Index(line, "#"); idx > 0 {
				line = strings.TrimRight(line[:idx], " \t")
			}
		}

		result = append(result, line)
	}

	return result
}

func parseYAML(content string, ignoreOrder bool) (string, error) {
	var data interface{}
	err := yaml.Unmarshal([]byte(content), &data)
	if err != nil {
		return "", err
	}

	if ignoreOrder {
		data = sortMapRecursive(data)
	}

	out, err := yaml.Marshal(data)
	if err != nil {
		return "", err
	}

	return string(out), nil
}

func parseJSON(content string, ignoreOrder bool) (string, error) {
	var data interface{}
	err := json.Unmarshal([]byte(content), &data)
	if err != nil {
		return "", err
	}

	if ignoreOrder {
		data = sortMapRecursive(data)
	}

	var out []byte
	out, err = json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}

	return string(out), nil
}

func sortMapRecursive(obj interface{}) interface{} {
	switch v := obj.(type) {
	case map[string]interface{}:
		result := make(map[string]interface{})
		for k, val := range v {
			result[k] = sortMapRecursive(val)
		}
		return result
	case map[interface{}]interface{}:
		result := make(map[string]interface{})
		for k, val := range v {
			result[fmt.Sprint(k)] = sortMapRecursive(val)
		}
		return result
	case []interface{}:
		result := make([]interface{}, len(v))
		for i, val := range v {
			result[i] = sortMapRecursive(val)
		}
		return result
	default:
		return v
	}
}

func compareFiles(file1, file2 string, config Config) (bool, string) {
	fileType := detectType(file1)

	fmt.Printf("%s文件类型: %s%s\n", ColorCyan, fileType, ColorReset)
	fmt.Printf("%s文件1: %s%s\n", ColorCyan, file1, ColorReset)
	fmt.Printf("%s文件2: %s%s\n", ColorCyan, file2, ColorReset)
	fmt.Printf("%s%s%s\n\n", ColorCyan, strings.Repeat("=", 60), ColorReset)

	content1, err := readFile(file1)
	if err != nil {
		return false, fmt.Sprintf("%s错误: %v%s", ColorRed, err, ColorReset)
	}

	content2, err := readFile(file2)
	if err != nil {
		return false, fmt.Sprintf("%s错误: %v%s", ColorRed, err, ColorReset)
	}

	var str1, str2 string

	switch fileType {
	case "yaml":
		str1, err = parseYAML(content1, config.IgnoreOrder)
		if err != nil {
			fmt.Printf("%s警告: YAML解析失败，回退到文本对比: %v%s\n", ColorYellow, err, ColorReset)
			return compareText(content1, content2, file1, file2, fileType, config)
		}
		str2, err = parseYAML(content2, config.IgnoreOrder)
		if err != nil {
			fmt.Printf("%s警告: YAML解析失败，回退到文本对比: %v%s\n", ColorYellow, err, ColorReset)
			return compareText(content1, content2, file1, file2, fileType, config)
		}

	case "json":
		str1, err = parseJSON(content1, config.IgnoreOrder)
		if err != nil {
			fmt.Printf("%s警告: JSON解析失败，回退到文本对比: %v%s\n", ColorYellow, err, ColorReset)
			return compareText(content1, content2, file1, file2, fileType, config)
		}
		str2, err = parseJSON(content2, config.IgnoreOrder)
		if err != nil {
			fmt.Printf("%s警告: JSON解析失败，回退到文本对比: %v%s\n", ColorYellow, err, ColorReset)
			return compareText(content1, content2, file1, file2, fileType, config)
		}

	default:
		return compareText(content1, content2, file1, file2, fileType, config)
	}

	return unifiedDiff(str1, str2, file1, file2, config.Context)
}

func compareText(content1, content2, file1, file2, fileType string, config Config) (bool, string) {
	lines1 := normalizeText(content1, fileType, config)
	lines2 := normalizeText(content2, fileType, config)

	str1 := strings.Join(lines1, "\n")
	str2 := strings.Join(lines2, "\n")

	return unifiedDiff(str1, str2, file1, file2, config.Context)
}

func unifiedDiff(str1, str2, file1, file2 string, context int) (bool, string) {
	diff := difflib.UnifiedDiff{
		A:        difflib.SplitLines(str1),
		B:        difflib.SplitLines(str2),
		FromFile: file1,
		ToFile:   file2,
		Context:  context,
	}

	result, err := difflib.GetUnifiedDiffString(diff)
	if err != nil {
		return false, fmt.Sprintf("%s错误: %v%s", ColorRed, err, ColorReset)
	}

	if result == "" {
		return true, fmt.Sprintf("%s✓ 文件完全相同%s", ColorGreen, ColorReset)
	}

	// 彩色化输出
	colored := colorizeDiff(result)
	return false, colored
}

func colorizeDiff(diff string) string {
	var result strings.Builder
	scanner := bufio.NewScanner(strings.NewReader(diff))

	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "---") || strings.HasPrefix(line, "+++") {
			result.WriteString(ColorCyan + ColorBold + line + ColorReset + "\n")
		} else if strings.HasPrefix(line, "@@") {
			result.WriteString(ColorYellow + line + ColorReset + "\n")
		} else if strings.HasPrefix(line, "+") {
			result.WriteString(ColorGreen + line + ColorReset + "\n")
		} else if strings.HasPrefix(line, "-") {
			result.WriteString(ColorRed + line + ColorReset + "\n")
		} else {
			result.WriteString(line + "\n")
		}
	}

	return result.String()
}

func compareDirs(dir1, dir2, pattern string, config Config) {
	files1, _ := filepath.Glob(filepath.Join(dir1, pattern))
	files2, _ := filepath.Glob(filepath.Join(dir2, pattern))

	// 提取文件名
	names1 := make(map[string]bool)
	for _, f := range files1 {
		names1[filepath.Base(f)] = true
	}

	names2 := make(map[string]bool)
	for _, f := range files2 {
		names2[filepath.Base(f)] = true
	}

	// 找出共有文件
	var common []string
	for name := range names1 {
		if names2[name] {
			common = append(common, name)
		}
	}
	sort.Strings(common)

	fmt.Printf("\n%s对比共有文件 (%d 个):%s\n\n", ColorCyan, len(common), ColorReset)

	sameCount := 0
	diffCount := 0

	for _, name := range common {
		file1 := filepath.Join(dir1, name)
		file2 := filepath.Join(dir2, name)

		fmt.Printf("\n%s%s%s\n", ColorCyan, strings.Repeat("=", 60), ColorReset)
		fmt.Printf("%s对比: %s%s\n", ColorCyan, name, ColorReset)
		fmt.Printf("%s%s%s\n\n", ColorCyan, strings.Repeat("=", 60), ColorReset)

		same, diff := compareFiles(file1, file2, config)
		fmt.Println(diff)

		if same {
			sameCount++
		} else {
			diffCount++
		}
	}

	fmt.Printf("\n%s%s%s\n", ColorCyan, strings.Repeat("=", 60), ColorReset)
	fmt.Printf("%s汇总%s\n", ColorCyan, ColorReset)
	fmt.Printf("%s%s%s\n", ColorCyan, strings.Repeat("=", 60), ColorReset)
	fmt.Printf("%s相同: %d%s\n", ColorGreen, sameCount, ColorReset)
	fmt.Printf("%s差异: %d%s\n", ColorRed, diffCount, ColorReset)
	fmt.Printf("总计: %d\n", sameCount+diffCount)
}

/*
go.mod:

module config-differ

go 1.21

require (
	github.com/pmezard/go-difflib v1.0.0
	gopkg.in/yaml.v3 v3.0.1
)
*/
